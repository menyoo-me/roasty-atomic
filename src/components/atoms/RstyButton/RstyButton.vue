<template>
  <component
    :is="href ? 'a' : 'button'"
    :href="href ? (disabled ? '#' : href) : null"
    :type="href ? type : type != null ? type : 'button'"
    :class="[this.baseClass, this.variationClasses, this.blockClasses]"
    :disabled="disabled"
    :role="href === '#' ? 'button' : null"
  >
    <slot />
  </component>
</template>

<script>
export default {
  name: 'rsty-button',
  props: {
    /**
     * The html element used for the button.
     * `button, a`
     */
    el: {
      type: String,
      default: 'button',
      validator: (value) => value.match(/(button|a)/),
    },
    /**
     * When setting the button’s type to a link, use this option to give a href.
     */
    href: {
      type: String,
      default: null,
    },
    /**
     * Set the button’s type to “submit”.
     */
    type: {
      type: String,
      default: null,
      validator: (value) => value.match(/(null|submit|button)/),
    },
    /**
     * Style variation to give additional meaning.
     * `primary, secondary`
     */
    variation: {
      type: String,
      default: 'default',
      validator: (value) => value.match(/(default|primary|secondary|cancel)/),
    },
    /**
     * Set true or false to disable button or replace the href link to '#'.
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Makes the button fill 100% of the container
     */
    block: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    baseClass: [
      'button',
      'outline-none',
      'active:outline-none',
      'focus:outline-none',
      'py-2',
      'px-3',
      'border',
      'rounded',
      'items-center',
      'font-bold',
      'transition',
      'duration-50',
      'ease-in-out',
      'transform',
      'active:scale-98',
    ],
    buttonVariationClasses: {
      disabled: [
        'bg-gray-200',
        'border-gray-300',
        'text-gray-400',
        'cursor-default',
        'cursor-not-allowed',
        'active:scale-100',
        'button--disabled',
      ],
      default: ['bg-gray-100', 'border-gray-500', 'text-gray-600'],
      primary: [
        'bg-blue-500',
        'border-blue-500',
        'text-white',
        'hover:bg-blue-400',
        'hover:border-blue-400',
        'button--primary',
      ],
      secondary: [
        'bg-white',
        'border-blue-500',
        'text-blue-500',
        'hover:text-blue-600',
        'hover:border-blue-600',
        'hover:shadow-lg',
        'button--secondary',
      ],
      cancel: ['bg-white', 'border-white', 'text-blue-500', 'button--cancel'],
    },
  }),
  computed: {
    variationClasses() {
      if (this.disabled) return this.buttonVariationClasses.disabled
      return this.buttonVariationClasses[this.variation]
    },
    blockClasses() {
      if (this.block) return ['flex', 'justify-center', 'w-full']
      return ['inline-flex']
    },
  },
}
</script>
