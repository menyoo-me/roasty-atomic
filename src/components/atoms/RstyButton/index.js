import RstyButton from './RstyButton'

export default RstyButton
