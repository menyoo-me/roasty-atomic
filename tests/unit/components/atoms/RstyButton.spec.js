import { createFactory } from '@/factories'
import { mount } from '@vue/test-utils'
import RstyButton from '@/components/atoms/RstyButton'

const factory = createFactory(({ props, options }) => {
  return mount(RstyButton, {
    propsData: { ...props },
    ...options,
  })
})

describe('RstyButton.vue', () => {
  it('toMatchSnapshot', async () => {
    const wrapper = factory()
    expect(wrapper).toMatchSnapshot()
  })

  it('has default structure and classes', async () => {
    const wrapper = factory()

    expect(wrapper.element.tagName).toBe('BUTTON')
    expect(wrapper.attributes('type')).toBeDefined()
    expect(wrapper.attributes('type')).toBe('button')
    expect(wrapper.attributes('href')).not.toBeDefined()
    expect(wrapper.attributes('role')).not.toBeDefined()
    expect(wrapper.attributes('disabled')).not.toBeDefined()
    expect(wrapper.attributes('aria-pressed')).not.toBeDefined()
    expect(wrapper.attributes('autocomplete')).not.toBeDefined()
    expect(wrapper.attributes('tabindex')).not.toBeDefined()
    expect(wrapper).toMatchSnapshot()

    wrapper.destroy()
  })

  it('renders a link when href provided', async () => {
    const props = {
      href: '/foo/bar',
    }
    const wrapper = factory({ props })

    expect(wrapper.element.tagName).toBe('A')
    expect(wrapper.attributes('href')).toBeDefined()
    expect(wrapper.attributes('href')).toBe('/foo/bar')
    expect(wrapper.attributes('type')).not.toBeDefined()
    expect(wrapper.attributes('role')).not.toBeDefined()
    expect(wrapper.attributes('disabled')).not.toBeDefined()
    expect(wrapper.attributes('aria-pressed')).not.toBeDefined()
    expect(wrapper.attributes('autocomplete')).not.toBeDefined()
    expect(wrapper.attributes('tabindex')).not.toBeDefined()

    expect(wrapper).toMatchSnapshot()

    wrapper.destroy()
  })

  it('renders default slot content', async () => {
    const options = {
      slots: {
        default: '<span>foobar</span>',
      },
    }
    const wrapper = factory({ options })

    expect(wrapper.element.tagName).toBe('BUTTON')
    expect(wrapper.attributes('type')).toBeDefined()
    expect(wrapper.attributes('type')).toBe('button')
    expect(wrapper.find('span').exists()).toBe(true)
    expect(wrapper.text()).toBe('foobar')

    expect(wrapper).toMatchSnapshot()

    wrapper.destroy()
  })
})
