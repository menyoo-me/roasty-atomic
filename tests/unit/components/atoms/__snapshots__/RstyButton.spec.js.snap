// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RstyButton.vue button has attribute disabled when disabled set 1`] = `<button type="button" disabled="disabled" class=""></button>`;

exports[`RstyButton.vue has default structure and classes 1`] = `<button type="button" class=""></button>`;

exports[`RstyButton.vue link disabled should have href="#" and role="button" 1`] = `<a href="#" role="button" class=""></a>`;

exports[`RstyButton.vue link with href="#" should have role="button" 1`] = `<a href="#" role="button" class=""></a>`;

exports[`RstyButton.vue renders a link when href provided 1`] = `<a href="/foo/bar" class=""></a>`;

exports[`RstyButton.vue renders default slot content 1`] = `<button type="button" class=""><span>foobar</span></button>`;

exports[`RstyButton.vue toMatchSnapshot 1`] = `<button type="button" class=""></button>`;
